### PR TITLE
Better handling of alias analysis for aten::backward

### PR DESCRIPTION
**Pull request by smessmer**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #36683 Use overload names for aten::backward
* **#36522 Better handling of alias analysis for aten::backward**

Differential Revision: [D21004102](https://our.internmc.facebook.com/intern/diff/D21004102/)

### CODE REVIEW

File: torch/csrc/autograd/VariableTypeManual.cpp [hunk 1]
```diff
@@ -375,7 +375,7 @@ static auto registry = torch::RegisterOperators()
     .impl_unboxedOnlyKernel<decltype(VariableType::copy_), &VariableType::copy_>(DispatchKey::VariableTensorId))
   .op(torch::RegisterOperators::options()
     .schema("aten::backward(Tensor self, Tensor? gradient=None, bool keep_graph=False, bool create_graph=False) -> ()")
-    .aliasAnalysis(AliasAnalysisKind::FROM_SCHEMA)
+    .aliasAnalysis(AliasAnalysisKind::CONSERVATIVE)
```
Comment (bhosmer) [inline torch/csrc/autograd/VariableTypeManual.cpp:378]: Hmmm, I was expecting to see `Tensor(a!) self` in the schema. As given, won't the change to `CONSERVATIVE` change behavior?

File: torch/csrc/jit/runtime/operator.h [hunk 2]
```diff
@@ -149,20 +149,12 @@ struct TORCH_API Operator {
...
-      // that affect more than just its own arguments.
-      // TODO Find a better way to do this.
-      TORCH_CHECK(
-          alias_analysis == AliasAnalysisKind::FROM_SCHEMA ||
-              !schemaRef.hasAnyAliasInfo(),
-          "In operator registration: Tried to register operator ",
-          schemaRef,
-          " with aliasing information in the schema but without AliasAnalysisKind::FROM_SCHEMA.");
-    }
+    TORCH_CHECK(
```
Comment (bhosmer) [inline torch/csrc/jit/runtime/operator.h:152]: Similarly, I'm not sure why the assert dodge was necessary at all, given that the schema for backward didn't have aliasing info (unless I'm misreading)

**Comment by wanchaol** [comment]
Should we also remove the special casing on `aten::backward` in `aliasAnalysisKind()`?

File: torch/csrc/autograd/VariableTypeManual.cpp (hunk 1 above)
Comment (smessmer) [reply torch/csrc/autograd/VariableTypeManual.cpp:378]: CONSERVATIVE implies that every argument is mutable and aliased with every other argument. The `(a!)` was actually unnecessary.

Comment (smessmer) [reply torch/csrc/autograd/VariableTypeManual.cpp:378]: ah that's still part of the misexport. There is a location that did specify the `(a!)`. I now added its removal to the PR.

Comment (bhosmer) [reply torch/csrc/autograd/VariableTypeManual.cpp:378]: Can we not use the same string literal for this registration and equivalent (second) one in -[register_prim_ops_fulljit](https://github.com/pytorch/pytorch/pull/36522/files#diff-fb26a0de03180fc8778ce18261715d98R153)? Even now they're not quite the same :)

Comment (smessmer) [reply torch/csrc/autograd/VariableTypeManual.cpp:378]: fixed

**Comment by smessmer** [comment]
@wanchaol that's what this PR is doing right? Or do you mean some other `aliasAnalysisKind`?

File: torch/csrc/jit/runtime/operator.h (hunk 2 above)
Comment (albanD) [reply torch/csrc/jit/runtime/operator.h:152]: I think the (a!) was aliasing info no?

File: torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp [hunk 3]
```diff
@@ -150,7 +150,7 @@ RegisterOperators reg(
          },
          aliasAnalysisConservative()),
      Operator(
-         "aten::backward(Tensor(a!) self, Tensor? gradient=None, bool? retain_graph=None, bool create_graph=False) -> ()",
+         "aten::backward(Tensor self, Tensor? gradient=None, bool? retain_graph=None, bool create_graph=False) -> ()",
```
Comment (wanchaol) [inline torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:153]: I guess we could try to delete this overload here, as in VariableTypeManual we already this as an operator.

File: torch/csrc/jit/runtime/operator.h (hunk 2 above)
Comment (wanchaol) [reply torch/csrc/jit/runtime/operator.h:152]: yeah but this PR deleted `(a!)` from the schema and use conservative aliasing, so backward is not having any aliasing information i think.

**Review by wanchaol — COMMENTED** [review]
> @wanchaol that's what this PR is doing right? Or do you mean some other `aliasAnalysisKind`?

you are right, sorry I didn't pay close attention to it last time. This looks good to me, I am wondering if we could delete the jit operator as we registered the operator in VariableTypeManual. 

Edit: Feature request on operator registration, I think it's better for operator registration to have this functionality: If I registered the same operator twice (with the same operator name and schema), we should just throw error/warning. It's good to warn the user that we already have the same operator otherwise overload resolution might be confusing.

**Comment by smessmer** [comment]
superseded by superseded by https://github.com/pytorch/pytorch/pull/36750

**Comment by smessmer** [comment]
> Edit: Feature request on operator registration, I think it's better for operator registration to have this functionality: If I registered the same operator twice (with the same operator name and schema), we should just throw error/warning. It's good to warn the user that we already have the same operator otherwise overload resolution might be confusing.

agreed, I'm working on that on the side, but there's currently a lot of ops that do register themselves multiple times... :(

**Pull request closed without merging**